### ISSUE TITLE
add support for generating common functions (copyWith, to/fromMap, to/fromJson) 

### ISSUE THREAD

**Issue by pq** [dart-lang/sdk#48971]
Form a discussion on flutter-insiders:

> Some common things could be included too in the Generate... menu, like the copyWith method or the toMap, fromMap (from and to json)

Currently the `"Generate..."` menu produces this in IntelliJ

<img width="288" alt="image" src="https://user-images.githubusercontent.com/67586/167181814-d8221aa8-05d5-4c66-8bc1-c05fc6258451.png">

@DanTup: do we have similar support in VSCode?

(And are these current menu items implementing in IntelliJ or server-provided?)

/cc @stevemessick

**Comment by DanTup** [comment]
We don't have anything like this in VS Code right now (because it would be hard to do reliable in VS Code versus the server).  It does come up now and then though - it'd be awesome to support :-)

**Comment by bwilkerson** [comment]
Would these be appropriate for snippets, or would they be better as commands?

**Comment by DanTup** [comment]
Yeah, I think snippets (eg. code completion) could be a good place for them. If the goal is for them to be used in both editors, we'd need to add support for snippets to the original protocol.

It could also be useful to have a way to re-generate them if you add new fields (perhaps as an assist on the method?).

I also wonder if there's any overlap with [static metaprogramming](https://github.com/dart-lang/language/issues/1482) - some of the things above are specifically noted as use cases in that issue. I can see reasons for both (being able to have invisible generated code, or having them just visible in the code, or being able to generate a basic implementation to customise), but I wonder if there should be some relationship between them (for example, might it be useful to "inline" some code that was previously being generated with static meta programming to customise, and could/should these snippets be driven by that?).

**Comment by bwilkerson** [comment]
> If the goal is for them to be used in both editors, we'd need to add support for snippets to the original protocol.

@jwren We'll need to talk to Alex and get a clear direction for where we want this to go in IntelliJ, but if we do nothing else we could always return them as code completions the way we're doing in LSP.

> I also wonder if there's any overlap with static metaprogramming ...

Yes, absolutely. This is one of the use cases that static metaprogramming would support. That proposal has the advantage that the methods would automatically be kept up to date.

That said, Dart team hasn't yet decided whether static metaprogramming will be approved and supported, and even if it is, I expect it will take a while for it to be implemented. It might well be worthwhile for us to support some of the more common cases in the meantime.
